### ISSUE TITLE
Add CURL loss?

### ISSUE THREAD

**Issue by JohnGiorgi** [KevinMusgrave/pytorch-metric-learning#81]
I came across an interesting looking loss in this [paper](https://arxiv.org/abs/1902.09229) that the authors call [CURL](https://arxiv.org/abs/1902.09229) (not to be confused with this other [CURL](https://arxiv.org/abs/2004.04136)). I was wondering if this idea could be ported to the library? 

![Screen Shot 2020-04-30 at 10 27 34 PM](https://user-images.githubusercontent.com/8917831/80776993-cac83c00-8b31-11ea-8e5c-1f2cbbebb9d8.png)

The loss is given in equation 12 under section 6.3, it is used to improve performance over strong baselines in both image/text representation learning. It looks a little like the InfoNCE/NT-Xent losses, but there are two sums introduced over "blocks" of positives and "blocks" of negatives.

I thought about trying to implement it myself and open a pull request but I wanted to first get the repo maintainers' opinion on whether or not this would be possible, and if so what the best way to approach it might be. Thanks!

> I reached out to the authors [via their blog](http://disq.us/p/29008i9) to see if they have a implementation somewhere.

**Comment by KevinMusgrave** [comment]
I suppose the steps would be to mine pairs or triplet (or create the pair/triplet indices yourself). Then group pairs/triplets by anchor indices. So basically for each sample in the batch, you have all the positives and negatives associated with it. Given that, the equation shouldn't be difficult to implement. The loss could have two options, hinge and logistic (I think that's what the small L represents).

**Comment by JohnGiorgi** [comment]
Okay I understand, I ended up implementing it more or less as you suggested, but ended up just doing the averaging over my "blocks" of positives/negatives outside the loss function, then using the loss functions from this library as-is.

I think it might make more sense to implement CURL as a simple loss wrapper? Something that accepts blocks of positives/negatives and does the averaging for you before passing them on to the wrapped loss. No idea how to do this at the moment so I am going to go ahead and close the issue while I think about it.
